assets/src/components/ui/hover-text/index.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * Internal dependencies
+ */
+import './style.css';
+
+export const HOVER_TEXT_POSITION_TOP_LEFT = 'top-left';
+export const HOVER_TEXT_POSITION_TOP_RIGHT = 'top-right';
+export const HOVER_TEXT_POSITION_BOTTOM_LEFT = 'bottom-left';
+export const HOVER_TEXT_POSITION_BOTTOM_RIGHT = 'bottom-right';
+
+/**
+ * HoverText
+ * Adds a text element that is only displayed when the wrapped component (children) is hovered over
+ *
+ * @function
+ * @param {string} hoverText    The actual text to be displayed when the child entity is hovered over
+ * @param {string} htmlId       Used to generate an HTML "id" attribute.
+ * 								The description and "-hover-text" are appended.
+ * @param {string} htmlClass    Used to generate an HTML "class" attribute.
+ * 								The description and "-hover-text" are appended.
+ * @param {string} description  Two or three words (hyphenated) that describe the component
+ *                              that the hover text is being added to
+ * @param {string} position     Where the hover text appears relative to the wrapped component
+ * @param {string} children     The child entity that gets wrapped with the hover elements
+ * @return {string}             The child entity wrapped with the hover elements
+ */
+export const HoverText = ( { hoverText, htmlId, htmlClass, description, position, children } ) => {
+	position = position ?
+		position :
+		HOVER_TEXT_POSITION_TOP_RIGHT;
+	const pointerCharCodes = {};
+	pointerCharCodes[ HOVER_TEXT_POSITION_TOP_LEFT ] = 9701;
+	pointerCharCodes[ HOVER_TEXT_POSITION_TOP_RIGHT ] = 9700;
+	pointerCharCodes[ HOVER_TEXT_POSITION_BOTTOM_LEFT ] = 9698;
+	pointerCharCodes[ HOVER_TEXT_POSITION_BOTTOM_RIGHT ] = 9699;
+	return hoverText && (
+		<div
+			id={ `${ htmlId }-${ description }-hover-text` }
```

Comment:
So its okay to have potentially invalid html rendered?

Reply:
If an html id attribute is not needed then I would say don't bother adding it.

Reply:
> would actually be counter to our purposes.

What purposes would it be counter to?

Reply:
> It's really only needed if a dev wants to tweak the styling for a specific hover text in a component

Reply:
If that's the case, and you decide to keep this component,  I'd prefer it if you either:

- remove the id and developers can tweak styling by the `htmlClass` prop.  I don't see how having an id is necessary if they can provide a className.  If an id is necessary for styling then that is argument for it being unique.
OR
- wrap the component with `withInstanceId` and implement so the html id attribute is unique.

However, my preference is that you just use `PopOver` component from `@wordpress/components` because from what I've seen so far, I don't see the need for re-inventing the wheel here.

Reply:
> So its okay to have potentially invalid html rendered?

No you are making a strawman argument as that was not my point. A form input for example can still work correctly even if a developer fails to give it a unique id. That does not mean that invalid html is ok, it only means that nothing breaks as a result of that invalid html. With Gutenberg blocks, it is critical that some components have a unique id else the editor will not be able to target them properly and things will not work.

> I don't see how having an id is necessary if they can provide a className.

class names are not intended to be unique so if you have multiple elements that share the same class, then you need to start relying on more complex css to target a specific element, which may not always be possible.

> If an id is necessary for styling then that is argument for it being unique.

There is nothing preventing the `htmlId` prop here from being unique. It **CAN**  be unique simply by setting a unique value when adding the component to your app. This is no different than our PHP form system that generates ids for form elements in that it is up to the developer to ensure that the data they are inputting results in valid html. If i had two inputs in a row in a form and used "my-input" for the id for both inputs, then that would result in invalid html but it would NOT prevent the form from operating correctly in any other way, as long as I did not require those input ids for any kind of targeting (JS or CSS). Having a duplicate id for some hover text is not good, but it will not break anything. 

With Gutenberg blocks it is a totally different scenario because the editor can create multiple instances of a block dynamically, so I could add one, two, three... ten... one hundred... (one miiiillllliiiion) copies of the same block to a post, and in order for GB to be able to target each one of those instances separately, they all need to have an id that is at least partially dynamically generated. This can also apply to components being used within the block, especially if you are adding form elements to the Gutenberg sidebar, because again, those may need to have a unique id for targeting purposes. 

Here is the first sentence from the documentation for the `withInstanceId` HOC:

> Some components need to generate a unique id for each instance.

There is nothing in the documentation stating that every single component that implements an id must absolutely use the `withInstanceId` HOC in order to guarantee the uniqueness of that id. But **SOME** components do need to generate a unique id for each instance. To suggest that a non-critical component like some hover text needs to utilize the `withInstanceId` HOC in order to guarantee a totally unique id attribute is totally overboard. If it is that much of an issue, then the more sane approach would be to use the `withInstanceId` HOC on a more appropriate parent component higher up in the hierarchy of your app, and then pass that id down to any child elements and have them incorporate that into any ids that they set. In other words, there is no reason why the html id for some hover text could not incorporate a unique id that was passed down from a `withInstanceId` HOC that was used higher up in the app.

Reply:
I understand how `withInstanceId` works and its purpose.  I'm not making a straw man argument.  My point is that IF someone was to generate a bunch of these components without passing in a unique `htmlId` prop (and other props used for the id are the same), then the result would be invalid html rendered.  I'm not saying it won't work, I'm saying it would be _invalid_.  So my question was simply asking "So its okay to have **potentially** invalid html rendered?" (note I emphasized potentially here).  You're answer can simply be that you believe it is.

I still stand by my outlined preferences here.

Reply:
There is nothing stopping this html id prop from being unique, but the dev implementing it in their component would be responsible for actually generating a unique id in the first place, and that is no different than a bunch of our other systems that we already have in place (like our form system).

Again, the sane thing to do would be to use `withInstanceId` on a more appropriate parent component in any app that can be implemented more than once on the same page, and then pass that unique id down through the component hierarchy. In other words, only use the `withInstanceId HOC` for things that can have multiple instances generated dynamically. It's total overkill to employ it every single time you plan to implement an html id attribute.

Reply:
I understand that.  However, that's why one of the options I gave was:

> remove the id and developers can tweak styling by the htmlClass prop. I don't see how having an id is necessary if they can provide a className. If an id is necessary for styling then that is argument for it being unique.

Developers can pass along a unique className for the css class if they want.  In other words don't set an id if it's not necessary.

Now you didn't mention it, but just in case its one possible reason  you are explicitly adding an id here (beyond for styling purposes), if the id is something you're adding so the element can be targeted by js, it'd be better if you wrap the component with `React.forwardRef` and set a `ref` on the element you want targeted as that's the recommended way to integrate custom binding to rendered elements.